test/test_storage.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2019-2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import gc
+import sys
+import tempfile
+from textwrap import dedent
+
+from ops import (
+    framework,
+    storage,
+)
+from test.test_helpers import (
+    BaseTestCase,
+    fake_script,
+    fake_script_calls,
+)
+
+
+class StoragePermutations(abc.ABC):
+
+    @abc.abstractmethod
+    def create_framework(self) -> framework.Framework:
+        """Create a Framework that we can use to test the backend storage.
+        """
+        return NotImplemented
+
+    def test_save_and_load_snapshot(self):
+        f = self.create_framework()
+
+        class Sample(framework.Object):
+
+            def __init__(self, parent, key, content):
+                super().__init__(parent, key)
+                self.content = content
+
+            def snapshot(self):
+                return {'content': self.content}
+
+            def restore(self, snapshot):
+                self.__dict__.update(snapshot)
+
+        f.register_type(Sample, None, Sample.handle_kind)
+        content = {
+            'str': 'string',
+            'bytes': b'bytes',
+            'int': 1,
+            'float': 3.0,
+            'dict': {'a': 'b'},
+            'set': {'a', 'b'},
+            'list': [1, 2],
+        }
+        s = Sample(f, 'test', content)
+        handle = s.handle
+        f.save_snapshot(s)
+        del s
+        gc.collect()
+        res = f.load_snapshot(handle)
+        self.assertEqual(content, res.content)
+
+
+class TestSQLiteStorage(StoragePermutations, BaseTestCase):
+
+    def create_framework(self):
+        return framework.Framework(':memory:', None, None, None)
+
+
+class _MemoryStorageBackend:
```

Comment:
I had started prototyping a minimal 'state-set/get-delete' scripts that would write things out to a file and then read it back in again. However, 'fake_script' files *must* be bash scripts, so I didn't persue it very far. I could do it without fake_script and just drop python scripts into $PATH, or we can ignore the disk interaction.
I do think that interaction is most of the interesting bit of this patch, but I'm testing that directly w Test.*Backend.

Reply:
I'll be messing with (ie refactoring) fake_script as soon as #322 is landed, maybe it'll help?